<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · GBIO.jl</title><meta name="title" content="Home · GBIO.jl"/><meta property="og:title" content="Home · GBIO.jl"/><meta property="twitter:title" content="Home · GBIO.jl"/><meta name="description" content="Documentation for GBIO.jl."/><meta property="og:description" content="Documentation for GBIO.jl."/><meta property="twitter:description" content="Documentation for GBIO.jl."/><meta property="og:url" content="https://GenomicBreeding.github.io/GBIO.jl/"/><meta property="twitter:url" content="https://GenomicBreeding.github.io/GBIO.jl/"/><link rel="canonical" href="https://GenomicBreeding.github.io/GBIO.jl/"/><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="search_index.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>GBIO.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/GenomicBreeding/GBIO.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/GenomicBreeding/GBIO.jl/blob/main/docs/src/index.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="GBIO"><a class="docs-heading-anchor" href="#GBIO">GBIO</a><a id="GBIO-1"></a><a class="docs-heading-anchor-permalink" href="#GBIO" title="Permalink"></a></h1><p>Documentation for <a href="https://github.com/jeffersonfparil/GBIO.jl">GBIO</a>.</p><ul><li><a href="#GBIO.isfuzzymatch-Tuple{String, String}"><code>GBIO.isfuzzymatch</code></a></li><li><a href="#GBIO.levenshteindistance-Tuple{String, String}"><code>GBIO.levenshteindistance</code></a></li><li><a href="#GBIO.readJLD2-Union{Tuple{Type{T}}, Tuple{T}, Tuple{Type{T}, String}} where T&lt;:GBCore.AbstractGB"><code>GBIO.readJLD2</code></a></li><li><a href="#GBIO.readdelimited-Tuple{Type{GBCore.Genomes}}"><code>GBIO.readdelimited</code></a></li><li><a href="#GBIO.readdelimited-Tuple{Type{GBCore.Trials}}"><code>GBIO.readdelimited</code></a></li><li><a href="#GBIO.readdelimited-Tuple{Type{GBCore.Phenomes}}"><code>GBIO.readdelimited</code></a></li><li><a href="#GBIO.readvcf-Tuple{}"><code>GBIO.readvcf</code></a></li><li><a href="#GBIO.writeJLD2-Tuple{GBCore.AbstractGB}"><code>GBIO.writeJLD2</code></a></li><li><a href="#GBIO.writedelimited-Tuple{GBCore.Trials}"><code>GBIO.writedelimited</code></a></li><li><a href="#GBIO.writedelimited-Tuple{GBCore.Phenomes}"><code>GBIO.writedelimited</code></a></li><li><a href="#GBIO.writedelimited-Tuple{GBCore.Genomes}"><code>GBIO.writedelimited</code></a></li><li><a href="#GBIO.writevcf-Tuple{GBCore.Genomes}"><code>GBIO.writevcf</code></a></li></ul><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBIO.isfuzzymatch-Tuple{String, String}" href="#GBIO.isfuzzymatch-Tuple{String, String}"><code>GBIO.isfuzzymatch</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">isfuzzymatch(a::String, b::String; threshold::Float64=0.3)::Bool</code></pre><p>Fuzzy string matching using the Levenshtein distance, and a threshold as a fraction of the smaller string.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; isfuzzymatch(&quot;populations&quot;, &quot;populations&quot;)
true

julia&gt; isfuzzymatch(&quot;populations&quot;, &quot;poplation&quot;)
true

julia&gt; isfuzzymatch(&quot;populations&quot;, &quot;entry&quot;)
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBIO.jl/blob/06fe6965d79b14c0b344a51ceea7d4cfe32be9e0/src/fuzzy_matching.jl#L47-L63">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBIO.levenshteindistance-Tuple{String, String}" href="#GBIO.levenshteindistance-Tuple{String, String}"><code>GBIO.levenshteindistance</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">levenshteindistance(a::String, b::String)::Int64</code></pre><p>Calculate the Levenshtein distance between 2 strings. TODO: optimise as we only need to compute 2 rows at a time and do not need the full matrix</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; levenshteindistance(&quot;populations&quot;, &quot;populations&quot;)
0

julia&gt; levenshteindistance(&quot;populations&quot;, &quot;poplation&quot;)
2

julia&gt; levenshteindistance(&quot;populations&quot;, &quot;entry&quot;)
3</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBIO.jl/blob/06fe6965d79b14c0b344a51ceea7d4cfe32be9e0/src/fuzzy_matching.jl#L1-L18">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBIO.readJLD2-Union{Tuple{Type{T}}, Tuple{T}, Tuple{Type{T}, String}} where T&lt;:GBCore.AbstractGB" href="#GBIO.readJLD2-Union{Tuple{Type{T}}, Tuple{T}, Tuple{Type{T}, String}} where T&lt;:GBCore.AbstractGB"><code>GBIO.readJLD2</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">readJLD2(type::Type, fname::String = missing)::Type</code></pre><p>Load a core (<code>Genomes</code>, <code>Phenomes</code>, and <code>Trials</code>), simulation (<code>SimulatedEffects</code>), and model (<code>TEBV</code>) struct from a JLD2 file.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = GBCore.simulategenomes(n=2, sparsity=0.01, verbose=false);

julia&gt; fname = writeJLD2(genomes);

julia&gt; readJLD2(Genomes, fname) == genomes
true

julia&gt; phenomes = Phenomes(n=2, t=2); phenomes.entries = [&quot;entry_1&quot;, &quot;entry_2&quot;]; phenomes.traits = [&quot;trait_1&quot;, &quot;trait_2&quot;];

julia&gt; fname = writeJLD2(phenomes);

julia&gt; readJLD2(Phenomes, fname) == phenomes
true

julia&gt; trials = Trials(n=1, t=2); trials.entries = [&quot;entry_1&quot;];

julia&gt; fname = writeJLD2(trials);

julia&gt; readJLD2(Trials, fname) == trials
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBIO.jl/blob/06fe6965d79b14c0b344a51ceea7d4cfe32be9e0/src/reader.jl#L1-L29">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBIO.readdelimited-Tuple{Type{GBCore.Genomes}}" href="#GBIO.readdelimited-Tuple{Type{GBCore.Genomes}}"><code>GBIO.readdelimited</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">readdelimited(type::Type{Genomes}; fname::String, sep::String = &quot;\t&quot;)::Genomes</code></pre><p>Load a <code>Genomes</code> struct from a string-delimited (default=<code>\t</code>) file.  Each row corresponds to a locus-allele combination. The first 4 columns correspond to the chromosome, position, all alleles in the locus (delimited by <code>|</code>), and the specific allele. The subsequency columns refer to the samples, pools, entries or genotypes.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = GBCore.simulategenomes(n=10, sparsity=0.1, verbose=false);

julia&gt; fname = writedelimited(genomes);

julia&gt; genomes_reloaded = readdelimited(Genomes, fname=fname);

julia&gt; genomes == genomes_reloaded
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBIO.jl/blob/06fe6965d79b14c0b344a51ceea7d4cfe32be9e0/src/reader.jl#L44-L63">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBIO.readdelimited-Tuple{Type{GBCore.Phenomes}}" href="#GBIO.readdelimited-Tuple{Type{GBCore.Phenomes}}"><code>GBIO.readdelimited</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">readdelimited(type::Type{Phenomes}; fname::String, sep::String = &quot;\t&quot;)::Phenomes</code></pre><p>Load a <code>Phenomes</code> struct from a string-delimited (default=<code>\t</code>) file.  Each row corresponds to a locus-allele combination. The first 4 columns correspond to the chromosome, position, all alleles in the locus (delimited by <code>|</code>), and the specific allele. The subsequency columns refer to the samples, pools, entries or genotypes.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; phenomes = Phenomes(n=10, t=3); phenomes.entries = string.(&quot;entry_&quot;, 1:10); phenomes.populations .= &quot;pop1&quot;; phenomes.traits = [&quot;A&quot;, &quot;B&quot;, &quot;C&quot;]; phenomes.phenotypes = rand(10,3); phenomes.phenotypes[1,1] = missing; phenomes.mask .= true;

julia&gt; fname = writedelimited(phenomes);

julia&gt; phenomes_reloaded = readdelimited(Phenomes, fname=fname);

julia&gt; phenomes == phenomes_reloaded
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBIO.jl/blob/06fe6965d79b14c0b344a51ceea7d4cfe32be9e0/src/reader.jl#L266-L285">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBIO.readdelimited-Tuple{Type{GBCore.Trials}}" href="#GBIO.readdelimited-Tuple{Type{GBCore.Trials}}"><code>GBIO.readdelimited</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">readdelimited(type::Type{Trials}; fname::String, sep::String = &quot;\t&quot;)::Trials</code></pre><p>Load a <code>Trials</code> struct from a string-delimited (default=<code>\t</code>) file.  Each row corresponds to a locus-allele combination. The first 4 columns correspond to the chromosome, position, all alleles in the locus (delimited by <code>|</code>), and the specific allele. The subsequency columns refer to the samples, pools, entries or genotypes.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = GBCore.simulategenomes(n=10, verbose=false);

julia&gt; trials, _ = GBCore.simulatetrials(genomes=genomes, sparsity=0.1, verbose=false);

julia&gt; fname = writedelimited(trials);

julia&gt; trials_reloaded = readdelimited(Trials, fname=fname);

julia&gt; trials == trials_reloaded
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBIO.jl/blob/06fe6965d79b14c0b344a51ceea7d4cfe32be9e0/src/reader.jl#L445-L466">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBIO.readvcf-Tuple{}" href="#GBIO.readvcf-Tuple{}"><code>GBIO.readvcf</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">readvcf(;fname::String, field::String = &quot;any&quot;, verbose::Bool = false)::Genomes</code></pre><p>Load Genomes struct from vcf file</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = GBCore.simulategenomes(n=10, sparsity=0.1, verbose=false);

julia&gt; fname = writevcf(genomes);

julia&gt; genomes_reloaded = readvcf(fname=fname);

julia&gt; genomes.entries == genomes_reloaded.entries
true

julia&gt; dimensions(genomes) == dimensions(genomes_reloaded)
true

julia&gt; ismissing.(genomes.allele_frequencies) == ismissing.(genomes_reloaded.allele_frequencies)
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBIO.jl/blob/06fe6965d79b14c0b344a51ceea7d4cfe32be9e0/src/reader.jl#L619-L641">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBIO.writeJLD2-Tuple{GBCore.AbstractGB}" href="#GBIO.writeJLD2-Tuple{GBCore.AbstractGB}"><code>GBIO.writeJLD2</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">writeJLD2(A::Union{Genomes,Phenomes,Trials,SimulatedEffects}; fname::Union{Missing,String} = missing)::String</code></pre><p>Save core (<code>Genomes</code>, <code>Phenomes</code>, and <code>Trials</code>), simulation (<code>SimulatedEffects</code>), and model (<code>TEBV</code>) structs as JLD2,  a heirarchical data format version 5 (HDF5) - compatible format. Note that the extension name should be &#39;.jld2&#39;.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = GBCore.simulategenomes(n=2, verbose=false);

julia&gt; writeJLD2(genomes, fname=&quot;test_genomes.jld2&quot;)
&quot;test_genomes.jld2&quot;

julia&gt; genomes_reloaded = load(&quot;test_genomes.jld2&quot;);

julia&gt; genomes_reloaded[collect(keys(genomes_reloaded))[1]] == genomes
true

julia&gt; phenomes = Phenomes(n=2, t=2); phenomes.entries = [&quot;entry_1&quot;, &quot;entry_2&quot;]; phenomes.traits = [&quot;trait_1&quot;, &quot;trait_2&quot;];

julia&gt; writeJLD2(phenomes, fname=&quot;test_phenomes.jld2&quot;)
&quot;test_phenomes.jld2&quot;

julia&gt; phenomes_reloaded = load(&quot;test_phenomes.jld2&quot;);

julia&gt; phenomes_reloaded[collect(keys(phenomes_reloaded))[1]] == phenomes
true

julia&gt; trials = Trials(n=1, t=2); trials.entries = [&quot;entry_1&quot;];

julia&gt; writeJLD2(trials, fname=&quot;test_trials.jld2&quot;)
&quot;test_trials.jld2&quot;

julia&gt; trials_reloaded = load(&quot;test_trials.jld2&quot;);

julia&gt; trials_reloaded[collect(keys(trials_reloaded))[1]] == trials
true

julia&gt; simulated_effects = SimulatedEffects();

julia&gt; writeJLD2(simulated_effects, fname=&quot;test_simulated_effects.jld2&quot;)
&quot;test_simulated_effects.jld2&quot;

julia&gt; simulated_effects_reloaded = load(&quot;test_simulated_effects.jld2&quot;);

julia&gt; simulated_effects_reloaded[collect(keys(simulated_effects_reloaded))[1]] == simulated_effects
true

julia&gt; trials, _simulated_effects = GBCore.simulatetrials(genomes = GBCore.simulategenomes(n=10, verbose=false), n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=10, verbose=false);

julia&gt; tebv = analyse(trials, max_levels=50, verbose=false);

julia&gt; writeJLD2(tebv, fname=&quot;test_tebv.jld2&quot;)
&quot;test_tebv.jld2&quot;

julia&gt; tebv_reloaded = load(&quot;test_tebv.jld2&quot;);

julia&gt; tebv_reloaded[collect(keys(tebv_reloaded))[1]] == tebv
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBIO.jl/blob/06fe6965d79b14c0b344a51ceea7d4cfe32be9e0/src/writer.jl#L1-L62">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBIO.writedelimited-Tuple{GBCore.Genomes}" href="#GBIO.writedelimited-Tuple{GBCore.Genomes}"><code>GBIO.writedelimited</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">writedelimited(genomes::Genomes, sep::String = &quot;	&quot;, fname::Union{Missing,String} = missing)::String</code></pre><p>Save <code>Genomes</code> struct as a string-delimited (default=<code></code>) file. Each row corresponds to a locus-allele combination. The first 4 columns correspond to the chromosome, position, all alleles in the locus (delimited by <code>|</code>), and the specific allele. The subsequency columns refer to the samples, pools, entries or genotypes.</p><p><strong>Notes:</strong></p><ul><li>Extension name should be &#39;.tsv&#39;, &#39;.csv&#39;, or &#39;.txt&#39;.</li><li>Header lines and comments are prefixed by &#39;#&#39;.</li><li>There are 2 header lines prefixed by &#39;#&#39;, e.g.:<ul><li>header line 1: &quot;chrom,pos,all<em>alleles,allele,entry</em>1,entry_2&quot;</li><li>header line 2: &quot;chrom,pos,all<em>alleles,allele,population</em>1,population_1&quot;</li></ul></li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = GBCore.simulategenomes(n=2, verbose=false);

julia&gt; writedelimited(genomes, fname=&quot;test_genomes.tsv&quot;)
&quot;test_genomes.tsv&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBIO.jl/blob/06fe6965d79b14c0b344a51ceea7d4cfe32be9e0/src/writer.jl#L89-L111">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBIO.writedelimited-Tuple{GBCore.Phenomes}" href="#GBIO.writedelimited-Tuple{GBCore.Phenomes}"><code>GBIO.writedelimited</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">writedelimited(phenomes::Phenomes, sep::String = &quot;	&quot;, fname::Union{Missing,String} = missing)::String</code></pre><p>Save <code>Phenomes</code> struct as a string-delimited (default=<code></code>) file.  Each row corresponds to a samples, pools, entries or genotypes. The first 2 columns correspond to the entry and population names. The subsequency columns refer to the traits containing the phenotype values of each entry. Note that the extension name should be &#39;.tsv&#39;, &#39;.csv&#39;, or &#39;.txt&#39;.</p><p><strong>Notes:</strong></p><ul><li>Extension name should be &#39;.tsv&#39;, &#39;.csv&#39;, or &#39;.txt&#39;.</li><li>Header line and comments are prefixed by &#39;#&#39;.</li><li>There is 1 header line prefixed by &#39;#&#39;, e.g.: &quot;entry,population,trait<em>1,trait</em>2,trait_3&quot;</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; phenomes = Phenomes(n=2, t=2); phenomes.entries = [&quot;entry_1&quot;, &quot;entry_2&quot;]; phenomes.traits = [&quot;trait_1&quot;, &quot;trait_2&quot;];

julia&gt; writedelimited(phenomes, fname=&quot;test_phenomes.tsv&quot;)
&quot;test_phenomes.tsv&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBIO.jl/blob/06fe6965d79b14c0b344a51ceea7d4cfe32be9e0/src/writer.jl#L164-L185">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBIO.writedelimited-Tuple{GBCore.Trials}" href="#GBIO.writedelimited-Tuple{GBCore.Trials}"><code>GBIO.writedelimited</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">writedelimited(trials::Trials, sep::String = &quot;	&quot;, fname::Union{Missing,String} = missing)::String</code></pre><p>Save <code>Trials</code> struct as a string-delimited (default=<code></code>) file.  Each row corresponds to a samples, pools, entries or genotypes. The first 10 columns correspond to:</p><ol><li>years</li><li>seasons</li><li>harvests</li><li>sites</li><li>entries</li><li>populations</li><li>replications</li><li>blocks</li><li>rows</li><li>cols </li></ol><p>The subsequency columns refer to the traits containing the phenotype values.</p><p><strong>Notes:</strong></p><ul><li>Extension name should be &#39;.tsv&#39;, &#39;.csv&#39;, or &#39;.txt&#39;.</li><li>Header line and comments are prefixed by &#39;#&#39;.</li><li>There is 1 header line prefixed by &#39;#&#39;, e.g.: &quot;years,seasons,harvests, ..., trait<em>1,tratit</em>2,trait_3&quot;</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; trials = Trials(n=1, t=2); trials.years = [&quot;year_1&quot;]; trials.seasons = [&quot;season_1&quot;]; trials.harvests = [&quot;harvest_1&quot;]; trials.sites = [&quot;site_1&quot;]; trials.entries = [&quot;entry_1&quot;]; trials.populations = [&quot;population_1&quot;]; trials.replications = [&quot;replication_1&quot;]; trials.blocks = [&quot;block_1&quot;]; trials.rows = [&quot;row_1&quot;]; trials.cols = [&quot;col_1&quot;]; trials.traits = [&quot;trait_1&quot;, &quot;trait_2&quot;];

julia&gt; writedelimited(trials, fname=&quot;test_trials.tsv&quot;)
&quot;test_trials.tsv&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBIO.jl/blob/06fe6965d79b14c0b344a51ceea7d4cfe32be9e0/src/writer.jl#L236-L268">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBIO.writevcf-Tuple{GBCore.Genomes}" href="#GBIO.writevcf-Tuple{GBCore.Genomes}"><code>GBIO.writevcf</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">writevcf(genomes::Genomes; fname::Union{Missing,String}, ploidy::Int64=0, max_depth::Int64=100, n_decimal_places::Int64=4)::String</code></pre><p>Save <code>Genomes</code> struct as a variant call format (VCF version 4.2) file.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes_1 = GBCore.simulategenomes(n=2, verbose=false);

julia&gt; writevcf(genomes_1, fname=&quot;test_genomes_1.vcf&quot;)
&quot;test_genomes_1.vcf&quot;

julia&gt; genomes_2 = GBCore.simulategenomes(n=2, n_alleles=3, verbose=false);

julia&gt; genomes_2.allele_frequencies = round.(genomes_2.allele_frequencies .* 4) ./ 4;

julia&gt; writevcf(genomes_2, fname=&quot;test_genomes_2.vcf&quot;, ploidy=4)
&quot;test_genomes_2.vcf&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBIO.jl/blob/06fe6965d79b14c0b344a51ceea7d4cfe32be9e0/src/writer.jl#L339-L358">source</a></section></article></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.0 on <span class="colophon-date" title="Wednesday 8 January 2025 03:03">Wednesday 8 January 2025</span>. Using Julia version 1.11.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
